### ISSUE TITLE
ASan does not report passing a non-null-terminated char array to std::ostream << const char*

### ISSUE THREAD

**Issue by b1tflyyyy** [llvm/llvm-project#198201]
## Summary

I would like to clarify whether this is expected behavior or a bug / limitation in AddressSanitizer.

The program below passes a non-null-terminated character array to `std::ostream << const char*`. My understanding is that this is undefined behavior, because `operator<<(std::ostream&, const char*)` expects a null-terminated C string.

## Code

```cpp
#include <iostream>

struct Test
{
    static constexpr char kString[] = { '1', '2', '3', '4', '5', '6', '7', '8' };
};

int main()
{
    std::cout << Test::kString << '\n';
    return 0;
}
```

## Build Command

I compiled it with:

```sh
clang++ main.cpp \
  -std=c++20 -Wall -Wextra -Wpedantic \
  -g -O0 -fno-omit-frame-pointer \
  -fsanitize=address,leak,undefined
```

## Actual Behavior

At runtime, the output looks consistent with reading past the end of `kString`:

```
12345678��
```

However, I do not get an AddressSanitizer report.

## Notes

The same behavior is reproducible on Compiler Explorer: https://godbolt.org/z/cGa66Y3vc

In the generated assembly, I do not see any obvious ASan redzone around `Test::kString`:

```asm
Test::kString:
        .ascii  "12345678"
```

For comparison, ASan does report the error for this variant compiled with the same compiler and flags:
Compiler Explorer link for this variant: https://godbolt.org/z/KaoP113rW

```cpp
#include <iostream>

int main()
{
    static constexpr char kString[] = { '1', '2', '3', '4', '5', '6', '7', '8' };
    std::cout << kString << '\n';
    
    return 0;
}
```

That version produces an `AddressSanitizer: global-buffer-overflow` report:

```text
READ of size 9 at 0x... thread T0
    #0 ... in strlen
    #1 ... in std::operator<<(std::basic_ostream<char, std::char_traits<char>>&, char const*)
    #2 ... in main /app/example.cpp:6:15

0x... is located 0 bytes after global variable 'main::kString' defined in '/app/example.cpp:5' ... of size 8
SUMMARY: AddressSanitizer: global-buffer-overflow ... in std::operator<<(std::basic_ostream<char, std::char_traits<char>>&, char const*)
...
Global redzone:          f9
```

This difference is what makes me think I may be missing something about how ASan handles these two cases.

## Question

1. Is it expected that ASan does not diagnose this case?
2. If this is a known limitation, is there a recommended way to catch such cases?

**Comment by thurstond** [comment]
Thanks for the report! It's an ASan bug / limitation.

Suppose we had another file, `test.cpp` that contained an identical definition of `struct Test`. It's legal to compile `main.cpp` (from your example) and `test.cpp` together, even though there are two identical definitions of `struct Test` [*]; the linker will silently deduplicate them. To make this work, the compiler annotated each instance of the struct as "comdat" (common data).

In your original test case, the struct in `main.cpp` was still annotated as comdat just in case it gets linked with something like `test.cpp`. When ASan encounters a comdat global, it assumes there might be multiple copies of the global; since it doesn't know which copy will take precedence, it doesn't instrument any of them.

There's a TODO in ASan:
```
  // For non-COFF targets, only instrument globals known to be defined by this
  // TU.
  // FIXME: We can instrument comdat globals on ELF if we are using the
  // GC-friendly metadata scheme.
  if (!TargetTriple.isOSBinFormatCOFF()) {
    if (!G->hasExactDefinition() || G->hasComdat())
      return false;
```

-----

[*] https://en.cppreference.com/cpp/language/definition

"There can be more than one definition in a program of each of the following: class type ... as all following conditions are satisfied:

* Each definition appears in a different translation unit.
...
* Each definition consists of the same sequence of tokens (typically, appears in the same header)."

**Comment by b1tflyyyy** [comment]
Thanks for the explanation!
